src/main/java/teammates/ui/webapi/UpdateFeedbackSessionAction.java

```diff
@@ -103,21 +108,93 @@ public JsonResult execute() throws InvalidHttpRequestBodyException {
+        deadlinesToRevoke.keySet().removeAll(newDeadlines.keySet());
+
+        deadlinesToRevoke.keySet().forEach(email ->
+                logic.deleteDeadlineExtension(courseId, session.getFeedbackSessionName(), email, areInstructors));
+
+        // Create deadline extensions
+        Map<String, Instant> deadlinesToCreate = new HashMap<>(newDeadlines);
+        deadlinesToCreate.keySet().removeAll(oldDeadlines.keySet());
+
+        deadlinesToCreate.entrySet()
+                .stream()
+                .map(entry ->
+                    DeadlineExtensionAttributes
+                            .builder(courseId, session.getFeedbackSessionName(), entry.getKey(), areInstructors)
+                            .withEndTime(entry.getValue())
+                            .build())
+                .forEach(deadlineExtension -> {
+                    try {
+                        logic.createDeadlineExtension(deadlineExtension);
+                    } catch (InvalidParametersException | EntityAlreadyExistsException e) {
+                        log.severe("Unexpected error while creating deadline extension", e);
+                    }
+                });
+
+        // Update deadline extensions
+        Map<String, Instant> deadlinesToUpdate = new HashMap<>(newDeadlines);
+        deadlinesToUpdate = deadlinesToUpdate.entrySet().stream()
+                .filter(entry -> oldDeadlines.containsKey(entry.getKey())
+                        && !entry.getValue().equals(oldDeadlines.get(entry.getKey())))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        deadlinesToUpdate.entrySet()
+                .stream()
+                .map(entry -> DeadlineExtensionAttributes
+                            .updateOptionsBuilder(
+                                    courseId, session.getFeedbackSessionName(), entry.getKey(), areInstructors)
+                            .withEndTime(entry.getValue())
+                            .build())
+                .forEach(updateOptions -> {
+                    try {
+                        logic.updateDeadlineExtension(updateOptions);
+                    } catch (InvalidParametersException | EntityDoesNotExistException e) {
+                        log.severe("Unexpected error while updating deadline extension", e);
+                    }
+                });
+
+        if (notifyUsers) {
+            CourseAttributes course = logic.getCourse(courseId);
+            emailsToSend.addAll(emailGenerator
+                    .generateDeadlineRevokedEmails(course, session, deadlinesToRevoke, areInstructors));
+            emailsToSend.addAll(emailGenerator
+                    .generateDeadlineGivenEmails(course, session, deadlinesToCreate, areInstructors));
+            emailsToSend.addAll(emailGenerator
+                    .generateDeadlineUpdatedEmails(course, session, deadlinesToUpdate, oldDeadlines, areInstructors));
+        }
```

Comment:
Is it idiomatic to add to a `List` passed in, or to return it as an output (and let the caller to decide whether to add to the output list if any)?

Reply:
I just thought it'll be slightly more performant if we pass the list by reference (so that we do not have to create a new list then add all the contents to another list). Do you think this optimization is premature/unwarranted?

Reply:
nvm, I do think it is cleaner. I will do it your way.

Reply:
Premature indeed haha... the gain is probably marginal until we are sure from jmeter that it is a significant portion of the code path. We can easily investigate and optimize this part later on.